src/util/ProjectIO.as

```diff
@@ -207,48 +207,113 @@ public class ProjectIO {
 		}
 	}
 
+	// Load all images in all costumes from their image data, then call whenDone.
 	public function decodeAllImages(objList:Array, whenDone:Function, fail:Function = null):void {
-		// Load all images in all costumes from their image data, then call whenDone.
-		function imageDecoded():void {
-			for each (var o:* in imageDict) {
-				if (o == 'loading...') return; // not yet finished loading
+		// This should be called on success or failure of each image
+		function imageDone():void {
+			if (--numImagesToDecode == 0) {
```

Comment:
Ah... just to be clear, I made this change because I wanted to stop storing the magic string 'loading...' in the dictionary which usually contains image data (`BitmapData` or `SVGElement`), and using that to check whether each image is done. The alleged performance benefit was a side-effect :)

The alleged performance benefit: the old code would check every image in `imageDict` for the string `'loading...'` each time any image in `imageDict` finished for a total of O(n^2) string checks. The new code just decrements the integer and checks that against zero.

This form of count-based "is everybody done?" pattern is my usual, but only because I haven't encountered any method that I like better. I'm certainly open to suggestions. For example, some of the SVG code keeps two separate counts (one for submitted requests and one for finished requests) and declares the overall batch to be done when the two counts are equal.